### DISCUSSION TITLE
Renovate reopens previously closed PRs after a couple of months

### DISCUSSION THREAD

**Discussion by HenrikPoulsen** [renovatebot/renovate#19413] [Request Help]
Internally we have a package which has not been updated for over a year.
One of our monorepos do not want the latest version of this package so the PR has been closed since a long time.
However after a couple of months (and several hundred PRs later) renovate suddenly creates a new PR for the same dependency.
Same dependency, target version, same exact branch name, everything.
Renovate does always add the "Renovate Ignore Notification" after they are closed, but they end up being revived anyway.

Is there some known limitation/issue around this?
Does Renovate only check the X latests PRs or something and once enough PRs have been opened since it cant find the closed ones anymore and choses to reopen it?

**Comment by rarkins** [comment]
Yes, I think there's a limit to how many closed PRs we fetch. Are you GHE or GLS?

**Reply by HenrikPoulsen** [reply]
GHE. Is it something we can configure?

**Reply by rarkins** [reply]
I believe the relevant logic is here: https://github.com/renovatebot/renovate/blob/main/lib/modules/platform/github/pr.ts

@zharinov do you recall if GitHub's GraphQL API imposes the upper limit, or is it us?

**Reply by zharinov** [reply]
It's not GraphQL as far as I can see and I'm not sure why we can stop our iteration

**Reply by rarkins** [reply]
Do you mean that our code is intended to iterate until every PR is fetched?

**Reply by zharinov** [reply]
Yes, I think so

**Reply by rarkins** [reply]
@HenrikPoulsen you should check out this log message: https://github.com/renovatebot/renovate/blob/f37733dcb1cd4610418605b41cca81a27892a7eb/lib/modules/platform/github/pr.ts#L103-L110

Then see if you can validate the count is what you expect based on checking the Renovate user's PRs manually

**Reply by HenrikPoulsen** [reply]
I ran it with debug on the repo in question just now:
```
DEBUG: getPrList success (repository=redacted, branch=renovate/actions-codeowners-validator-digest)
       "pullsTotal": 128,
       "requestsTotal": 1,
       "apiQuotaAffected": true
```
The repo in question currently has 248 open PRs and 7058 closed ones.
I debugged the code and checked
https://github.com/renovatebot/renovate/blob/f37733dcb1cd4610418605b41cca81a27892a7eb/lib/modules/platform/github/pr.ts#L93-L94

both of those variables are set to true (and there is a valid `next` in the link header). Then it comes to this:
https://github.com/renovatebot/renovate/blob/f37733dcb1cd4610418605b41cca81a27892a7eb/lib/modules/platform/github/pr.ts#L96-L98
which sets `needNextPageFetch` to false and it stops fetching more PRs. Is this intended behavior?

**Reply by rarkins** [reply]
It should only be fetching its own PRs. Can you check how many (open and closed) are authored by your bot account?

BTW this logic is relevant to your last point: https://github.com/renovatebot/renovate/blob/f37733dcb1cd4610418605b41cca81a27892a7eb/lib/modules/platform/github/pr.ts#L66-L70

**Reply by HenrikPoulsen** [reply]
There is 66 open and 876 closed PRs authored by the renovate github app we have.
I took a look at what this contains
https://github.com/renovatebot/renovate/blob/f37733dcb1cd4610418605b41cca81a27892a7eb/lib/modules/platform/github/pr.ts#L75
And the res.body only returns the first 1000 PRs, it filters out the PRs that don't belong to the bot, which leaves 128 PRs.
The PR from the closed dependencies then most likely end up somewhere after those first 1000 hits.

I can understand that there's an upper cap, since you can end up with a crazy amount of PRs in a repo, but you end up in situations like this where renovate doesn't know that the PR has been previously closed since it's not in the 1000 most recently updated PRs anymore.
I assume you would need to use the gql api instead so you can filter on the author right? Since it doesn't look like the rest api supports filtering on the author

**Reply by rarkins** [reply]
For some reason I was thinking we use GraphQL for that PR list already.

Anyway if it's REST, you can try this env: https://github.com/renovatebot/renovate/blob/f37733dcb1cd4610418605b41cca81a27892a7eb/lib/util/http/github.ts#L315
